masschallenge/impact-api
# AC-5343: Build Environment Deployment Targets

to test this first create a new directory and then clone all three repos into it. 
```
mkdir modular-models-epic
cd modular-models-epic
git clone git@github.com:masschallenge/impact-api.git
git clone git@github.com:masschallenge/accelerate.git
git clone git@github.com:masschallenge/django-accelerator.git
```
then check out the AC-5343 branches on impact and accelerate. Also checkout the modular-models-epic branch on django-accelerator
```
cd accelerate && git checkout AC-5343
cd ../impact-api && git checkout AC-5343
cd ../django-accelerator && git checkout modular-models-epic
```

then run ```make build``` in the impact-api directory.

## testing the make release command

to test the release command:

1. cd into the impact-api project and create an empty commit via:
```git commit --allow-empty -m "test commit"```

    the release command will only create a release tag if there are changes since the last release. 

2. now run  ```make release``` to create a release in all 3 projects.

3. there should be a running travis build for the tagged release. once this completes - make note of the tag that was generated and move on to the next section.

## testing a deploy

1. to test a deploy, find a test opsworks stack and set the "custom cookbooks" branch to AC-5343:
<img width="839" alt="screen shot 2018-03-19 at 3 07 03 pm" src="https://user-images.githubusercontent.com/196425/37616962-f098aeb2-2b87-11e8-9dc6-84cc56c35184.png">

2. add the "impact_deploy::default" recipe to the "deploy" recipes section (OpsWorks -> [Stack] -> Layers -> Recipes) before the opsworks_deploy_python recipe:
<img width="958" alt="screen shot 2018-03-19 at 4 30 30 pm" src="https://user-images.githubusercontent.com/196425/37620573-ff31e5f0-2b92-11e8-810c-0e9e49c4f3d5.png">


3. set the following settings in the stack's  custom json - you can borrow test values for this from the [test3 instance](https://console.aws.amazon.com/opsworks/home?region=us-east-1#/stack/d4a32a10-b655-44f1-9e2c-262885c56176/show):

    IMPACT_ENVIRONMENT=test
    DOCKER_REGISTRY=997386117669.dkr.ecr.us-east-1.amazonaws.com
    ECS_ACCESS_KEY_ID=(get it from test3)
    ECS_SECRET_ACCESS_KEY=(get it from test3)

4. set the mc app's deploy branch/tag to the release tag that was created in the last section.

5. in ECS make note of the currently deployed task definition's image tag. The "test" cluster is the one that we have specified via the IMPACT_ENVIRONMENT setting. make note of the currently deployed tag (ECS -> [Cluster] -> Services -> click link under "Task Defin...", scroll down on resulting page):
<img width="1205" alt="screen shot 2018-03-19 at 3 17 35 pm" src="https://user-images.githubusercontent.com/196425/37617201-c392c208-2b88-11e8-80ea-eae1a9a2505f.png">

https://console.aws.amazon.com/ecs/home?region=us-east-1#/clusters/test/services

6. deploy the opsworks stack and wait for it to finish.

7. go back to the "test" cluster in ECS and look at the currently running task definition. note that the deployed image tag should have changed.
